Some PVS-Studio warning fixes (part 4)

dimhotepus opened the pull request:
Some missed warning fixes from PVS-Studio static analyzer tool.

dimhotepus committed:
[perf] Pass large (12 bytes) args by reference

dimhotepus committed:
[flow] Use static_assert instead of array size hacks

dimhotepus committed:
[flow] Ensure local variable doesn't hide member one

Easier to read, less probability of incorrect var mistake.

dimhotepus committed:
[init] Initialize vertex count to safe default

Set to UINT_MAX to catch uninitialized access alter (if smb doesn't call init()).

dimhotepus committed:
[init] Initialize class member to safe default

dimhotepus committed:
[flow] Ensure random generator evaluation order

* * has higher priority than >>. Make intention explicit by enclosing in ().
* Make const method const.
* Do not mix signed / unsigned types.
* Pass arguments of size <= cpu register size by value, since reference kills performance.

dimhotepus committed:
[dupl] Remove duplicated random generator

dimhotepus committed:
[flow] Passed by ref small value never changes, use pass by value

It makes no sense to pass value by reference if callee doesn't modify referenced object. It also incurs performance penalty for small (<= CPU register size) types. as generated callee code should derefence passed argument first to access referenced object.

dimhotepus committed:
[perf] Move assign to vectors if source vector is not used anymore

Also wrapped source vectors definitions in block to prevent their usage after move.

dimhotepus committed:
[init] Initialize class member to safe default

dimhotepus committed:
[init] Initialize class members in ctor parameter list

dimhotepus committed:
[init] Initialize class members to safe defaults

dimhotepus committed:
[init] Set members in ctor initializer list

dimhotepus committed:
[init] Initialize class members to safe defaults

dimhotepus committed:
[init] Set class members in ctor initialization list

dimhotepus committed:
[overflow] Prevent out of buffer write

j can reach sizeof(buffer) here, so buffer[j] can write value to one after buffer ends slot.

dimhotepus committed:
[free] Prevent null pointer dereference

dimhotepus committed:
[flow] Read profile sex correclty (now females are happier)

dimhotepus committed:
[flow] Do not dereference null pointer

dimhotepus committed:
[overflow] Prevent out of buffer write

j can reach sizeof(buffer), so writing to buffer[j] is UB.

dimhotepus committed:
[macros] Close variadic list after usage

dimhotepus committed:
[init] Always initialize rcode to prevent UB on access to

dimhotepus committed:
[flow] Use freopen return stream

On error freopen closes passed fp->stdio_stream and returns null, on success - returns new reopened file stream. So we need to store eopened stream or null in fp->stdio_stream.

Xottab-DUTY commented on src/xrUICore/ListWnd/UIListItemEx.cpp:
```diff
@@ -11,11 +11,10 @@
     //.	InitTexture("ui\\hud_map_point");
-    //.	SetStretchTexture(true);
-    m_dwSelectionColor = color_argb(200, 95, 82, 74);
+    //.	SetStretchTexture(true);    
```
Occasional indentation on the right side
```suggestion
    //.	SetStretchTexture(true);
```

dimhotepus replied:
Done.

Xottab-DUTY commented on src/xrUICore/Lines/UILine.cpp:
```diff
@@ -23,9 +23,8 @@ CUILine::~CUILine()
 }
 
 CUILine::CUILine(const CUILine& other)
+    : m_subLines(other.m_subLines), m_tmpLine(NULL) 
```
```suggestion
    : m_subLines(other.m_subLines), m_tmpLine(nullptr) 
```

dimhotepus replied:
Done.

Xottab-DUTY commented on src/xrPhysics/Geometry.cpp:
```diff
@@ -590,7 +588,7 @@ void CSphereGeom::set_build_position(const Fvector& ref_point)
 }
 
-CCylinderGeom::CCylinderGeom(const Fcylinder& cyl) { m_cylinder = cyl; }
+CCylinderGeom::CCylinderGeom(const Fcylinder& cyl) : m_cylinder(cyl) {}
```
```suggestion
CCylinderGeom::CCylinderGeom(const Fcylinder& cyl) : m_cylinder(cyl) { }
```

dimhotepus replied:
Done

Xottab-DUTY commented on src/xrPhysics/Geometry.cpp:
```diff
@@ -559,7 +557,7 @@ void CBoxGeom::set_build_position(const Fvector& ref_point)
 }
 
-CSphereGeom::CSphereGeom(const Fsphere& sphere) { m_sphere = sphere; }
+CSphereGeom::CSphereGeom(const Fsphere& sphere) : m_sphere(sphere) {}
```
```suggestion
CSphereGeom::CSphereGeom(const Fsphere& sphere) : m_sphere(sphere) { }
```

dimhotepus replied:
Done.

Xottab-DUTY commented on src/xrPhysics/Geometry.cpp:
```diff
@@ -405,7 +403,7 @@ void CODEGeom::destroy()
 }
 
-CBoxGeom::CBoxGeom(const Fobb& box) { m_box = box; }
+CBoxGeom::CBoxGeom(const Fobb& box) : m_box(box) {}
```
```suggestion
CBoxGeom::CBoxGeom(const Fobb& box) : m_box(box) { }
```

dimhotepus replied:
Done

Xottab-DUTY commented on src/xrGame/ui/UIRankIndicator.cpp:
```diff
@@ -3,7 +3,7 @@
 #include "xrUICore/Static/UIStatic.h"
 
-CUIRankIndicator::CUIRankIndicator() { m_current = u8(-1); }
+CUIRankIndicator::CUIRankIndicator() : m_current(u8(-1)) {}
```
```suggestion
CUIRankIndicator::CUIRankIndicator() : m_current(u8(-1)) { }
```

dimhotepus replied:
Done

Xottab-DUTY commented on src/xrGame/relation_registry_fights.cpp:
```diff
@@ -10,13 +10,9 @@
 //////////////////////////////////////////////////////////////////////////
 
 RELATION_REGISTRY::FIGHT_DATA::FIGHT_DATA()
+    : attacker(0xffffU), defender(0xffffU), total_hit(0), time(0), time_old(0),
```
```suggestion
    : attacker(u16(-1)), defender(u16(-1)), total_hit(0), time(0), time_old(0),
```

dimhotepus replied:
Done.

Xottab-DUTY commented on src/xrGame/Torch.cpp:
```diff
@@ -30,23 +30,18 @@ static const float OPTIMIZATION_DISTANCE = 100.f;
+    : fBrightness(1.f), lanim(nullptr), guid_bone(BI_NONE), m_delta_h(0),
+      m_switched_on(false), light_render(GEnv.Render->light_create()),
+      light_omni(GEnv.Render->light_create()), glow_render(GEnv.Render->glow_create()),
+      m_bNightVisionEnabled(false), m_bNightVisionOn(false), m_night_vision(nullptr)
```
Changed formatting.
```suggestion
    : fBrightness(1.f), lanim(nullptr), guid_bone(BI_NONE),
      m_delta_h(0), m_switched_on(false),
      light_render(GEnv.Render->light_create()),
      light_omni  (GEnv.Render->light_create()),
      glow_render (GEnv.Render->glow_create()),
      m_bNightVisionEnabled(false), m_bNightVisionOn(false), m_night_vision(nullptr)
```

dimhotepus replied:
Done.

Xottab-DUTY commented on src/xr_3da/AccessibilityShortcuts.hpp:
```diff
@@ -12,17 +12,10 @@ class AccessibilityShortcuts
 public:
     AccessibilityShortcuts()
+        : screensaverState(false),
+          stickyKeys{sizeof(stickyKeys)}, filterKeys{sizeof(filterKeys)}, toggleKeys{sizeof(toggleKeys)},
```
```suggestion
          stickyKeys{ sizeof(stickyKeys) }, filterKeys{ sizeof(filterKeys) }, toggleKeys{ sizeof(toggleKeys) },
```

dimhotepus replied:
Done.

Xottab-DUTY commented on src/xrNetServer/NET_Client.cpp:
```diff
@@ -330,20 +330,13 @@ void IPureClient::_Recieve(const void* data, u32 data_size, u32 /*param*/)
+    : device_timer(timer), NET(nullptr), net_Address_device(nullptr), net_Address_server(nullptr),
 #ifdef CONFIG_PROFILE_LOCKS
-    , net_csEnumeration(MUTEX_PROFILE_ID(IPureClient::net_csEnumeration))
+      , net_csEnumeration(MUTEX_PROFILE_ID(IPureClient::net_csEnumeration)),
```
```suggestion
    , net_csEnumeration(MUTEX_PROFILE_ID(IPureClient::net_csEnumeration)),
```

dimhotepus replied:
Done

Xottab-DUTY commented on src/xrD3D9-Null/IDirect3DIndexBuffer9.cpp:
```diff
@@ -8,23 +8,13 @@ const GUID DECLSPEC_SELECTANY IID_IDirect3DIndexBuffer9;
-    : m_refCount(0)
+    : m_refCount(0), m_pIDirect3DDevice9(pIDirect3DDevice9)
+//#ifdef D3D_DEBUG_INFO
+      , Name(nullptr), Length(iLength), Usage(iUsage), Format(iFormat), Pool(iPool),
```
```suggestion
    , Name(nullptr), Length(iLength), Usage(iUsage), Format(iFormat), Pool(iPool),
```

dimhotepus replied:
Done

Xottab-DUTY reviewed (commented):
Some changes require some more analysis, so I didn't finished the review yet

dimhotepus committed:
[flow] Fix Female profile loading

In gpiSaveDiskProfile we save client sex as follows:

if(profile->cache->sex == GP_MALE)
  fprintf(fp, "sex=Male\n");
if(profile->cache->sex == GP_FEMALE)
  fprintf(fp, "sex=Female\n");
if(profile->cache->sex == GP_PAT)
  fprintf(fp, "sex=Pat\n");

So to load client sex correctly we should inspect values at zero
index:

if(strcmp(key, "sex") == 0)
{
  if(toupper(value[0]) == 'M')
    infoCache.sex = GP_MALE;
  else if(toupper(value[0]) == 'F')
    infoCache.sex = GP_FEMALE;
  else // Assume Pat, not safe, but still.
    infoCache.sex = GP_PAT;
}

Xottab-DUTY commented on src/xrCore/LocatorAPI.cpp:
```diff
@@ -612,8 +612,7 @@ void CLocatorAPI::ProcessOne(pcstr path, const _finddata_t& entry)
             return;
         if (0 == xr_strcmp(entry.name, ".."))
             return;
-        if (path[xr_strlen(path) - 1] != _DELIMITER || path[xr_strlen(path) - 1] != '/')
```
@eagleivg, do you remember why you have added this check?

Xottab-DUTY replied:
Engine filesystem has strong requirements on path correctness.
```
Path can be delimited only by a single delimiter.
Delimiters cannot be mixed. (it's either \ or /, but not both)
Folder path must always end with delimiter.

Correct path on Windows: folder\\      (ends by single backslash only)
Correct path on Windows: folder\\file  (delimited by single backslash only)

Correct path on Linux:   folder/       (ends by single forward slash only)
Correct path on Linux:   folder/file   (delimited by single forward slash only)

Wrong path on Windows:   folder
Wrong path on Windows:   folder/file
Wrong path on Windows:   folder//file
Wrong path on Windows:   folder\\\\file

Wrong path on Linux:     folder
Wrong path on Linux:     folder//file
Wrong path on Linux:	 folder\\file
Wrong path on Linux:	 folder\\\\file
etc...
```

dimhotepus replied:
So always adding delimiter for directory looks good.
I remember when we replaced logical operator with `&&`, engine fails to load some files, so looks like condition is really expected to be `true`.

Xottab-DUTY replied:
Not exactly. It looks like this condition is always true on Windows, but Linux filesystem (not the engine one! I mean OS internals) can add a delimiter before we enter this function. Removing this check will result in a double delimiter.

dimhotepus replied:
Hm, for me it looks independent of OS:
```
if (path[xr_strlen(path) - 1] != _DELIMITER || path[xr_strlen(path) - 1] != '/')
```
to simplify reasoning, replace `path[xr_strlen(path) - 1]` with `A`:
```
if (A != _DELIMITER || A != '/')
```
replace `_DELIMITER` with `2` and '/' with `3` (`_DELIMITER` now is '\\'):
```
if (A != 2 || A != 3)
```
so this condition is always `true`, independent of `A` value.

Do you mean `_DELIMITER` is generated and OS-dependent (`/` for Unix-like or Mac)?
If so, may be replace condition with:
```
#if defined(_WIN32)
  xr_strcat(N, DELIMITER);
#else
  if (A != _DELIMITER)
    xr_strcat(N, DELIMITER);
#endif

?

dimhotepus replied:
> Not exactly. It looks like this condition is always true on Windows, but Linux filesystem (not the engine one! I mean OS internals) can add a delimiter before we enter this function. Removing this check will result in a double delimiter.

Removed this commit :)

Xottab-DUTY commented on src/xrCore/LocatorAPI.cpp:
```diff
@@ -1037,7 +1037,7 @@ FileStatus CLocatorAPI::exist(pcstr fn, FSType fsType /*= FSType::Virtual*/)
     {
         struct stat buffer;
-        buffer.st_size;
+        [[maybe_unused]] const auto& st_size = buffer.st_size;
```
If think we don't even need this :)

dimhotepus replied:
Done

Xottab-DUTY commented on src/xrAICore/Navigation/vertex_allocator_fixed_inline.h:
```diff
@@ -15,7 +15,7 @@
 
 TEMPLATE_SPECIALIZATION
-inline CFixedVertexAllocator::CDataStorage() { m_vertices.resize(ReserveSize); }
+inline CFixedVertexAllocator::CDataStorage() : m_vertex_count(-1) { m_vertices.resize(ReserveSize); }
```
```suggestion
inline CFixedVertexAllocator::CDataStorage() : m_vertex_count(u32(-1)) { m_vertices.resize(ReserveSize); }
```

dimhotepus replied:
Done

Xottab-DUTY commented on src/xrAICore/Navigation/level_graph.cpp:
```diff
@@ -11,6 +11,7 @@
 #include "xrEngine/profiler.h"
 
 CLevelGraph::CLevelGraph(const char* fileName)
+    : m_level_id(GameGraph::_LEVEL_ID(-1))
```
We can leave this change in this pull request.

dimhotepus replied:
Well, done).

Xottab-DUTY replied:
@dimhotepus, but you didn't moved **other** changes to this file and `problem_solver_inline.h` to another pull request

Xottab-DUTY reviewed (commented):
> Sorry for a huge PR

Not so big :)

> may be i can help by splitting suspicious or hard to guess changes into distinct PRs?

Yes! For now it is probably better to move the changes to `problem_solver_inline.h` and `level_graph.cpp` to another pull request. All other files are checked except these two...

Xottab-DUTY commented on src/xrAICore/Navigation/level_graph.cpp:
```diff
@@ -218,7 +219,7 @@ u32 CLevelGraph::vertex_id(const Fvector& position) const
             {
                 // so, new node is under the specified position
-                if (position.y - _y < position.y - y)
+                if (_y > y)
```
This change is not explained, so I can't understand why it's changed.

dimhotepus replied:
I've tried to simplify the expressions, consider:
```
if (position.y - _y < position.y - y) {
```

assume `position.y` is `A`:

```
const auto A = position.y;
if (A - _y < A - y) {
```

then we can simplify expression using algebra rules (subtract `A` from both sides):

```
if (- _y < - y) {
```

then multiply both sides on `-1`, which reverts expression sign (consider `1 < 2`, than `-1 > -2`):

```
if (-1 * -_y > -1 * -y) {
```

and, finally, simplify expression by partial evaluation:

```
if (_y > y) {
```

dimhotepus replied:
@Xottab-DUTY If change is too suspicious, and explanation is not well, i can remove the change, ok?

Xottab-DUTY replied:
@dimhotepus the explanation is very well. That's just me being too neat/careful. I want to see the values in-game, but I won't be able to do that till New Year. Sorry for taking too long on this PR.

So, yes, it would be better to move this particular change to another pull request. Since everything is reviewed except this, the PR can be merged straightforward if we remove this.
Also, if you can, consider to just remove the commit 553f8ab12f1a107a6c6edee2ee854ce594e0471d via rebase instead of making a reverting commit. Thank you!

dimhotepus replied:
Not a problem, thank you for review and help! 

I'll remove this one https://github.com/OpenXRay/xray-16/commit/553f8ab12f1a107a6c6edee2ee854ce594e0471d as you suggested.
